arkouda/history.py

```diff
@@ -0,0 +1,113 @@
+import readline
+from IPython.core.history import HistoryAccessor  # type: ignore
+from typing import List, Optional
+
+
+class HistoryRetriever:
+    """
+    HistoryRetriever is an abstract base class that defines the retrieve method signature
+    and implements _filter_arkouda_command
+    """
+    def _filter_arkouda_command(self, command: str, filter_string: str = 'ak') -> Optional[str]:
+        """
+        Returns command string if the filter string is in the command and the
+        command is not generate_history. Otherwise, returns None
+        """
+        return command if (filter_string in command and 'generate_history' not in command) else None
+
+    def retrieve(self, command_filter: Optional[str] = None,
+                 num_commands: Optional[int] = None) -> List[str]:
+        """
+        Generates list of commands executed within a Python REPL shell, Jupyter notebook,
+        or IPython notebook, with an optional command filter and number of commands to return.
+
+        Parameters
+        ----------
+        num_commands : int
+            The number of commands from history to retrieve
+        command_filter : str
+            String containing characters used to select a subset of command history.
+
+        Returns
+        -------
+        List[str]
+            A list of commands from the Python shell, Jupyter notebook, or IPython notebook
+        """
+        raise NotImplementedError("Derived classes must implement retrieve")
+
+
+class ShellHistoryRetriever(HistoryRetriever):
+    """
+    ShellHistoryRetriever implements the retrieve method to get command history from the
+    Python REPL shell.
+    """
+    def retrieve(self, command_filter: Optional[str] = None,
+                 num_commands: Optional[int] = None) -> List[str]:
+        """
+        Generates list of commands executed within the a Python REPL shell, with an
+        optional command filter and number of commands to return.
+
+        Parameters
+        ----------
+        num_commands : int
+            The number of commands from history to retrieve
+        command_filter : str
+            String containing characters used to select a subset of command history.
+
+        Returns
+        -------
+        List[str]
+            A list of commands from the Python shell, Jupyter notebook, or IPython notebook
+        """
+        length_of_history = readline.get_current_history_length()
+        num_to_return = num_commands if num_commands else length_of_history
+
+        if command_filter:
+            return [readline.get_history_item(i + 1) for i in range(length_of_history)
+                    if self._filter_arkouda_command(readline.get_history_item(i + 1),
+                                                    command_filter)][-num_to_return:]
```

Comment:
This is likely not super important at all since I imagine the history is relatively small. 

It looks like we are running the filter on every element in the history before whittling it down to `num_to_return`. I'm wondering if we'll ever hit a situation where the history is quite long and the user wants only a few things if this could take longer than necessary. 

Using a while loop instead would prevent this, but it probably doesn't make a big difference in actual performance

Reply:
@pierce314159 yeah, good point, but I figured the length of history is gonna be small enough that this will not be noticeable performance hit. Also, can't use while loop in list comprehension.